test/Microsoft.AspNetCore.Razor.Test.Common/Language/Legacy/SyntaxTreeNodeWriter.cs

```diff
@@ -0,0 +1,231 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+
+namespace Microsoft.AspNetCore.Razor.Language.Legacy
+{
+    internal class SyntaxTreeNodeWriter : ParserVisitor
+    {
+        private readonly TextWriter _writer;
+
+        public int Depth { get; set; }
+
+        public SyntaxTreeNodeWriter(TextWriter writer)
+        {
+            _writer = writer;
+        }
+
+        public override void VisitDefault(Block block)
+        {
+            WriteBlock(block);
+        }
+
+        public override void VisitDefault(Span span)
+        {
+            WriteSpan(span);
+        }
+
+        public override void VisitTagHelperBlock(TagHelperChunkGenerator chunkGenerator, Block block)
+        {
+            WriteBlock(block);
+            
+            if (block is TagHelperBlock tagHelperBlock)
+            {
+                // TODO: Should I add TagHelperBinding info?
```

Comment:
Probably smart to hash it and represent it as some value so we know if it changes. No need for the info to be verbose.

Reply:
I added a list of descriptor names.